src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java

```diff
@@ -0,0 +1,471 @@
+ * created <i>once</i> and used for all rows in the result set. This approach has several
+ * advantages:
+ *
+ * <ol>
+ *
+ * <li>The Dialect must use the column metadata to determine <i>how</i> to convert the values to the
+ * correct type so that the converted values can be added to the {@link Struct}. The Dialect can do
+ * this once for each result set, rather than having to do the logic for every row.</li>
+ *
+ * <li>Although most dialects will create converter functions that merely use one of the ResultSet
+ * methods to get the correct value, implementations can easily create converter functions that have
+ * more complex logic.</li>
+ *
+ * <li>If the Dialect can return a null converter when a field cannot be converted and should be
+ * skipped entirely for all rows. This more efficient and explicit than a more conventional approach
+ * that calls the dialect to convert each value of the result set.</li>
+ *
+ * <li>This functional approach makes it very easy for the connector's task to process each row of
+ * the result set by calling each of the {@link ColumnConverter} functions that were created
+ * once.</li>
+ *
+ * </ol>
+ *
+ * <p>This functional approach is less common with Java 7 because it requires create lots of
+ * anonymous implementations and is therefore more verbose. However, the same approach is far more
+ * common in Java 8 and using lambdas requires very little code. This will make it easy to convert
+ * to Java 8 in the near future.
+ *
+ * <h2>Packaging and Deploying</h2> The dialect framework uses Java's {@link
+ * java.util.ServiceLoader} mechanism to find and automatically register all {@link
+ * DatabaseDialectProvider} implementations on the classpath. Don't forget to include in your JAR
+ * file a {@code META-INF/services/io.confluent.connect.jdbc.dialect .DatabaseDialectProvider} file
+ * that contains the fully-qualified name of your implementation class (or one class per line if
+ * providing multiple implementations).
+ *
+ * <p>If the Dialect implementation is included in this project, the class and the service provider
+ * file will automatically be included in the JDBC Source connector's JAR file. However, it is also
+ * possible to create an implementation outside of this project and to package the dialect (s) and
+ * service provider file in a JAR file, and to then include that JAR file inside the JDBC
+ * Connector's plugin directory within a Connect installation.
+ *
+ * <p>This discovery and registration process uses DEBUG messages to report the {@link
+ * DatabaseDialectProvider} classes that are found and registered. If you have difficulties getting
+ * the connector to find and register your dialect implementation classes, check that your JARs have
+ * the service provider file and your JAR is included in the JDBC connector's plugin directory.
+ */
+public interface DatabaseDialect extends ConnectionProvider {
```

Comment:
This interface defines methods for all interaction with a database via the JDBC driver. Unfortunately, there are a lot of methods. Some methods may seem strange because they could be implemented by directly using the JDBC driver, but there are lots of different driver implementations and unfortunately all of them have their own quirks.

When you review, please look at the signatures of the public methods, since these will become part of the public API.